Language/Java/Transformer.hs

```diff
@@ -0,0 +1,391 @@
+          (analyze newScope classBody)
+  analyze scope recordDecl@(RecordDecl src modifiers ident typeParams recordFieldDecls refTypes classBody) =
+    let newScope = addFields scope recordDecl
+     in RecordDecl
+          src
+          (map (analyze newScope) modifiers)
+          ident
+          typeParams
+          recordFieldDecls
+          refTypes
+          (analyze newScope classBody)
+  analyze scope enumDecl@(EnumDecl src modifiers ident refTypes enumBody) =
+    let newScope = addFields scope enumDecl
+     in EnumDecl
+          src
+          (map (analyze newScope) modifiers)
+          ident
+          refTypes
+          (analyze newScope enumBody)
+
+instance Transform MemberDecl where
+  analyze scope (FieldDecl span modifiers type_ varDecls) = FieldDecl span (map (analyze scope) modifiers) type_ (map (analyze scope) varDecls)
+  analyze scope methodDecl@(MethodDecl span modifiers typeParams mbType ident formalParams exceptionTypes mbExp methodBody) =
+    MethodDecl
+      span
+      (map (analyze scope) modifiers)
+      typeParams
+      mbType
+      ident
+      (map (analyze scope) formalParams)
+      exceptionTypes
+      (fmap (analyze scope) mbExp)
+      (analyze (addFormalParams scope (methodParameters methodDecl)) methodBody)
+  analyze scope constructorDecl@(ConstructorDecl span modifiers typeParams ident formalParams exceptionTypes constructorBody) =
+    ConstructorDecl
+      span
+      (map (analyze scope) modifiers)
+      typeParams
+      ident
+      (map (analyze scope) formalParams)
+      exceptionTypes
+      (analyze (addFormalParams scope (methodParameters constructorDecl)) constructorBody)
+  analyze scope (MemberClassDecl classDecl) = MemberClassDecl (analyze scope classDecl)
+  analyze scope (MemberInterfaceDecl interfaceDecl) = MemberInterfaceDecl (analyze scope interfaceDecl)
+
+instance Transform Catch where
+  analyze scope catch@(Catch formalParam block) =
+    Catch
+      (analyze scope formalParam)
+      (analyze (addFormalParams scope (exceptionParameter catch)) block)
+
+instance Transform Block where
+  analyze scope block@(Block blockstmts) = Block (map (analyze (addLocalVars scope (localVarsInBlock block))) blockstmts)
+
+instance Transform Stmt where
+  analyze scope (StmtBlock block) = StmtBlock (analyze scope block)
+  analyze scope (IfThen span exp stmt) = IfThen span (analyze scope exp) (analyze scope stmt)
+  analyze scope (IfThenElse span exp stmt1 stmt2) = IfThenElse span (analyze scope exp) (analyze scope stmt1) (analyze scope stmt2)
+  analyze scope (While span exp stmt) = While span (analyze scope exp) (analyze scope stmt)
+  analyze scope basicFor@(BasicFor span mbForInit mbExp mbExps stmt) =
+    let newScope = addLocalVars scope (localVarsInStmt basicFor)
+     in BasicFor
+          span
+          (fmap (analyze scope) mbForInit)
+          (fmap (analyze newScope) mbExp)
+          (fmap (map (analyze newScope)) mbExps)
+          (analyze newScope stmt)
```

Comment:
Auch hier finde ich es eher unglücklich, dass du das gesamte `Stmt` and `localVarsInStmt` übergibst, du weißt hier doch schon viel genauer, wo du suchen musst.

Reply:
Diente auch hier um nur eine Funktion für alle benutzen zukönnen.